src/kbmod/search/raw_image_eigen.h

```diff
@@ -0,0 +1,250 @@
+#ifndef RAWIMAGEEIGEN_H_
+#define RAWIMAGEEIGEN_H_
+
+
+#include <vector>
+#include <fitsio.h>
+#include <float.h>
+#include <iostream>
+#include <stdexcept>
+#include <Eigen/Core>
+
+#include "common.h"
+#include "psf.h"
+#include "pydocs/raw_image_docs.h"
+
+
+namespace search {
+
+  struct Index {
+    unsigned i;
+    unsigned j;
+
+    Index(unsigned idx_i, unsigned idx_j)
+      : i(idx_i), j(idx_j) {}
+
+    Index(int x, int y)
+      : i(x), j(y) {}
+
+    Index(float x, float y)
+      : i(floor(x)), j(floor(y)) {}
+
+    Index(float x, float y, bool floor)
+      : i(static_cast<unsigned>(x)), j(static_cast<int>(y)) {}
+
+    //Index(struct Point *p)
+    //  : x(floor(p->x)), y(floor(p->y)) {}
+
+    bool is_within(const unsigned width, const unsigned height) const{
```

Comment:
But how are the floats cast? Sometimes we floor it, sometimes we cast it.

I couldn't really get the same results for a while using other packages. Partly pixel-to-coordinate conventions were to blame. With a pixel grid that starts at `-0.5`, negative half-floats are still zeroth index. Is `Index(0, 0)`, `Point(0, 0)` or `Point(0.5, 0.5)`? Behavior suggests `Index(0, 0)` is `Point(0, 0)` and pixels stretch `0-->1` in real value coordinates. Astropy uses `-0.5` to `0.5` with origin in `top-left` and index convention `(y, x)`. I believe Rubin labels the center of `lower-left` pixel with `Index(0, 0)`, it's pixels span half-integer values (I think implying again `(-0.5, -0.5)` but it isn't explicitly stated) but they use `(x, y)` index ordering (so col, row). Eigen is firmly in the `(y, x)` (i.e. row, col) camp as is NumPy, and indexes top-left as origin. 

Figuring that out made it so that I could repeat our results with other code. Notably this is what I found somewhat confusing:
for an array
```
1 1
1 0
```
sampled in 10 points, so `{(0.0, 0.0}, (0.0, 0.1) .. (1, 0.9), (1, 1)` for pixel `[0,0]` we got 
```
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   1.00   
1.00   1.00   1.00   1.00   1.00   1.00   0.99   0.98   0.97   0.96   
1.00   1.00   1.00   1.00   1.00   1.00   0.98   0.96   0.94   0.92   
1.00   1.00   1.00   1.00   1.00   1.00   0.97   0.94   0.91   0.88   
1.00   1.00   1.00   1.00   1.00   1.00   0.96   0.92   0.88   0.84  
```

which didn't seem quite right, but it is what is expected for interpolation to return. There is a bit of weirdness that can sometimes happen when we hit the exact border because of the `floor(x+0.5)+0.5` that is used is slightly different than `floor(x+0.49)+0.5`. I was eventually able to find a way to replicate the result by:
```
from scipy import interpolate
xold = [0.5, 1.5]
yold = [0.5, 1.5]
pixels = np.array([[1., 1.], [1., 0]])
finterp = interpolate.interp2d(xold, yold, pixels)
xnew = np.arange(0, 1.1, 0.1)
ynew = np.arange(0, 1.1, 0.1)
interp_pixels = finterp(xnew, ynew)
print(interp_pixels)
```
resulting in
```
[[1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   1.  ]
 [1.   1.   1.   1.   1.   1.   0.99 0.98 0.97 0.96 0.95]
 [1.   1.   1.   1.   1.   1.   0.98 0.96 0.94 0.92 0.9 ]
 [1.   1.   1.   1.   1.   1.   0.97 0.94 0.91 0.88 0.85]
 [1.   1.   1.   1.   1.   1.   0.96 0.92 0.88 0.84 0.8 ]
 [1.   1.   1.   1.   1.   1.   0.95 0.9  0.85 0.8  0.75]]
```
Of course the problem now being that `interp2d` is being obsoleted and I can't seem to recover the same result via the [replacement functionality](https://scipy.github.io/devdocs/notebooks/interp_transition_guide.html). I'm not sure what I concluded, but it's something we should document more explicitly.

I'm still reasonably sure that the slight discrepancies mostly come from the somewhat inconsistent way we handle conversion of types not appropriate for indices into indices.

Reply:
As you note, KBMOD uses the convention where the pixel spans 0 -> 1. Point(0.5, 0.5) corresponds to the middle of pixel Index(0, 0).

Reply:
Agreed that we should be consistent about casting vs flooring. For the most part, it shouldn't matter. The pixel grid is defined to have only positive values, and casting + flooring are equivalent for positive values. The only place it would matter is at the edge. If a trajectory's predicted position falls right off the edge of the grid (-0.4, -0.3), a cast will pull it to (0, 0) while the floor will push it to (-1, -1).

Reply:
At a high level I think we can probably get rid of interpolated pixels altogether. We only use it for adding fake objects and for creating stamps for visualization.